### PR TITLE
#184651087 Implement Notifications for user to get updated on the status of the products

### PR DESCRIPTION
**Pull request by ishimwe5555**
#### What does this PR do?
Sends notifications to relevant users according to the status of their products
#### Description of Task to be completed?
Provided a product
GIVEN. the product lifecycle changes (added, removed, expired, bought, etc) then  the notification should be sent to the the right user
#### How should this be manually tested?
-Clone the repo and run npm run migrate:all

then login and put token in bearer
you can check your notifications via endpoint /notifications/all with GET method

#### Any background context you want to provide?
Notifications are being sent inApp and via email
#### What are the relevant pivotal tracker/Trello stories?
https://www.pivotaltracker.com/story/show/184651087
#### Screenshots (if appropriate)
n/a
#### Questions:

### CODE REVIEW

File: src/tests/notification.test.js [hunk 1]
```diff
@@ -0,0 +1,159 @@
+/* eslint-disable no-unused-expressions */
```
Comment (Pacific911) [inline src/tests/notification.test.js:1]: I think this eslint-disable is not needed, you can remove it

**Review by Pacific911 — CHANGES_REQUESTED** [review]

File: src/controllers/notification.controller.js [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+  });
+};
+
+// const NotificationAccess = async (req, res) => {
+//   res.status(200).json({
+//     code: 200,
+//     message: 'Access Granted',
+//     data: req.user,
+//   });
+// };
```
Comment (munezerobagira) [inline src/controllers/notification.controller.js:19]: why this comments

File: src/helpers/notifications.js [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+      socket.emit('all-notifications', myNotifications);
+    });
+    // Subscribe to new-notification event
+    notificationSystem.on('new-notification', (notification) => {
+      socket.emit('new-notification', notification);
+    });
+
+    // Disconnect event
+    socket.on('disconnect', () => {});
+  });
```
Comment (munezerobagira) [inline src/helpers/notifications.js:20]: we can use middleware instead setting the id 
https://socket.io/docs/v4/middlewares/

File: src/controllers/notification.controller.js (hunk 2 above)
Comment (ishimwe5555) [reply src/controllers/notification.controller.js:19]: Resolved. I mistakenly forgot to uncomment that while debugging.

Comment (ishimwe5555) [reply src/controllers/notification.controller.js:19]: Resolved. I mistakenly forgot to uncomment that while debugging.

File: src/helpers/notifications.js (hunk 3 above)
Comment (ishimwe5555) [reply src/helpers/notifications.js:20]: Thanks for the review, Useful, Let me put a hand into it.

Comment (ishimwe5555) [reply src/helpers/notifications.js]: Resolved

File: src/controllers/product.controller.js [hunk 4]
```diff
@@ -87,6 +88,14 @@ const addproduct = async (req, res) => {
...
         });
         await Promise.all(sendImage);
+
+        // await notificationUtils.createdProduct(req.user, productName);
+        // notificationServices.sendNotification(
+        //   userId,
+        //   'Product is created successfully',
+        //   'Product creation',
+        //   'low'
+        // );
```
Comment (princoo) [inline src/controllers/product.controller.js:98]: clean unnecessary comments

File: src/controllers/product.controller.js [hunk 5]
```diff
@@ -189,6 +198,13 @@ const updateOnadd = async (req, res) => {
           });
           newImages = await Promise.all(imageAdd);
         }
+        // await notificationUtils.updatedProduct(req.user, productName);
+        // notificationServices.sendNotification(
+        //   userId,
+        //   'Product is updated successfully',
+        //   'Product updated',
+        //   'low'
+        // );
```
Comment (princoo) [inline src/controllers/product.controller.js:207]: same here

File: src/index.js [hunk 6]
```diff
@@ -45,6 +51,10 @@ app.use('/', router);
 
 app.use(errorHandler);
 
-app.listen(PORT);
+const server = app.listen(PORT, () => {
+  console.log('Server started on port 3000');
+});
```
Comment (princoo) [inline src/index.js:56]: remove  this

**Review by princoo — CHANGES_REQUESTED** [review]

File: src/helpers/notifications.js [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+    }
+    const verifiedTOken = await redisClient.get(user.id);
+    if (!verifiedTOken) {
+      return false;
+    }
+    return user;
+  } catch (error) {
+    return false;
+  }
+}
```
Comment (Edcreation) [inline src/helpers/notifications.js:23]:
```suggestion
async function isValidAuthToken(token) {
  let user = null;
  try {
    if (token) {
      user = decodeToken(token);
      const verifiedToken = await redisClient.get(user.id);
      if (verifiedToken) {
        return user;
      }
    }
  } catch (error) {
    return false;
  }
  return false;
}

```

Comment (Edcreation) [reply src/helpers/notifications.js:23]: could refactor those returns;

**Review by Edcreation — CHANGES_REQUESTED** [review]

File: src/index.js (hunk 6 above)
Comment (ishimwe5555) [reply src/index.js:55]: Sure

File: src/helpers/notifications.js (hunk 7 above)
Comment (ishimwe5555) [reply src/helpers/notifications.js]: Sure did. Thanks for the review.

File: src/helpers/notifications.js [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+        return;
+      }
+      const { username, id } = user;
+      socket.id = id;
+      socket.username = username;
+      socket.emit('joined', { id: socket.id, username: socket.username });
+    });
+    socket.on('notification', async (msg) => {
+      console.log('*', msg);
+    });
```
Comment (Edcreation) [inline src/helpers/notifications.js:39]: This is not needed.

File: src/helpers/noti.js [hunk 9]
```diff
@@ -0,0 +1,7 @@
+const io = require('socket.io').listen(5000);
+
+io.sockets.on('connection', function (socket) {
+  socket.on('connection name', function (user) {
+    io.sockets.emit('new user', `${user.name} has joined.`);
+  });
+});
```
Comment (Edcreation) [inline src/helpers/noti.js:7]: This is also not need, even not tested its dropping coverage. If a message is needed it could be in that other file.

**Review by Edcreation — CHANGES_REQUESTED** [review]

File: src/helpers/noti.js (hunk 9 above)
Comment (ishimwe5555) [reply src/helpers/noti.js]: Sure, worked on.

**Review by AdrineUWERA — APPROVED** [review]

File: src/tests/socket.test.js [hunk 10]
```diff
@@ -0,0 +1,63 @@
+/* eslint-disable mocha/no-top-level-hooks */
```
Comment (Pacific911) [inline src/tests/socket.test.js:1]: it's better to remove this eslint-disable

**Review by Pacific911 — CHANGES_REQUESTED** [review]

**Review by Edcreation — APPROVED** [review]

File: src/tests/socket.test.js (hunk 10 above)
Comment (ishimwe5555) [reply src/tests/socket.test.js]: Yes sure, resolved.

**Review by Pacific911 — APPROVED** [review]

**Review by princoo — APPROVED** [review]

**Review by munezerobagira — APPROVED** [review]

**Pull request merged**